src/wasapi_backend/mod.rs

```diff
@@ -0,0 +1,363 @@
+mod stream_handle;
+use std::thread;
+
+use ringbuf::RingBuffer;
+use stream_handle::WasapiStreamHandle;
+use wasapi::initialize_sta;
+
+use crate::{
+    error::RunConfigError, AudioBackendOptions, AudioBufferStreamInfo, AudioDeviceConfigOptions,
+    AudioDeviceOptions, AudioDeviceStreamInfo, Backend, BackendStatus, ChannelLayout, DeviceID,
+    ProcessHandler, ProcessInfo, RainoutConfig, RunOptions, StreamHandle, StreamInfo,
+};
+
+fn empty_err<T, U>(res: Result<T, U>) -> Result<T, ()> {
+    match res {
+        Ok(res_inner) => Ok(res_inner),
+        Err(_) => Err(()),
+    }
+}
+
+fn run_config_err<T>(res: Result<T, Box<dyn std::error::Error>>) -> Result<T, RunConfigError> {
+    match res {
+        Ok(inner) => Ok(inner),
+        Err(err) => Err(RunConfigError::PlatformSpecific(err)),
+    }
+}
+
+fn get_device_ids(collection: &wasapi::DeviceCollection) -> Result<Vec<DeviceID>, ()> {
+    let mut result = Vec::new();
+
+    for i in 0..empty_err(collection.get_nbr_devices())? {
+        let device = empty_err(collection.get_device_at_index(i))?;
+
+        result.push(DeviceID {
+            name: empty_err(device.get_friendlyname())?,
+            identifier: Some(empty_err(device.get_id())?),
+        });
+    }
+
+    Ok(result)
+}
+
+pub fn enumerate_audio_backend() -> Result<AudioBackendOptions, ()> {
+    // TODO: No idea if this is cheap
+    empty_err(initialize_sta())?;
+
+    let input_devices = empty_err(wasapi::DeviceCollection::new(&wasapi::Direction::Capture))?;
+    let output_devices = empty_err(wasapi::DeviceCollection::new(&wasapi::Direction::Render))?;
+
+    let device_count =
+        empty_err(input_devices.get_nbr_devices())? + empty_err(output_devices.get_nbr_devices())?;
+
+    Ok(AudioBackendOptions {
+        backend: Backend::Wasapi,
+        version: None,
+        device_options: Some(AudioDeviceOptions::LinkedInOutDevice {
+            in_devices: get_device_ids(&input_devices)?,
+            out_devices: get_device_ids(&output_devices)?,
+        }),
+        status: match device_count {
+            0 => BackendStatus::NoDevices,
+            _ => BackendStatus::Running,
+        },
+    })
+}
+
+pub fn enumerate_audio_device(_device: &DeviceID) -> Result<AudioDeviceConfigOptions, ()> {
+    Ok(AudioDeviceConfigOptions {
+        // WASAPI supports more, but figuring out which is trial-and-
```

Comment:
Yeah. This is why we will enumerate each device separately instead of all at once, because using trial and error will be a bit slow.